core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -1951,12 +1955,27 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
     policyGroupsToRoles.toList.traverse {case (email, roles) => googleIamDao.addIamRoles(GoogleProject(googleProject.value), email, MemberType.Group, roles)} // addIamRoles logic includes retries
   }
 
-  private def getGoogleProjectNumber(googleProjectId: GoogleProjectId): Future[GoogleProjectNumber] = {
-    gcsDAO.getGoogleProject(googleProjectId).map { p => Option(p.getProjectNumber) match {
-        case None => throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadGateway, s"Failed to retrieve Google Project Number for Google Project ${googleProjectId}"))
-        case Some(longProjectNumber) => GoogleProjectNumber(longProjectNumber.toString)
-      }
-    }
+  // do all the google-y things in Cloud Resource Manager to reduce the number of calls made to google so we can avoid quota issues
+  private def setUpProjectInCloudResourceManagerAndGetGoogleProjectNumber(googleProjectId: GoogleProjectId, newLabels: Map[String, String]): Future[GoogleProjectNumber] = {
+
+    val validLabels = gcsDAO.labelSafeMap(newLabels)
+
+    for {
+      googleProject <- gcsDAO.getGoogleProject(googleProjectId)
+      googleProjectNumber = gcsDAO.getGoogleProjectNumberFromOption(googleProjectId, Option(GoogleProjectNumber(googleProject.getProjectNumber.toString)))
+
+      // RBS projects already come with some labels. In order to not lose those, we need to combine those existing labels with the new labels
+      existingLabels = googleProject.getLabels.asScala
+
+      updatedLabels = existingLabels ++ validLabels
+
+      // create a Project with fields that we want to update. Then send that to gcsDAO to update the project
+      googleProjectWithValuesToUpdate = new Project()
+        .setName(gcsDAO.googleProjectNameSafeString("Terra-managed Google project"))
```

Comment:
TOL: This seems like minimal information here, would we want some sort of identifier on which Terra billing project it was associated with? I just wonder if we can give some information to users at all. Although maybe it doesn't matter. Do we expect users to care about their google project?

If so, who does care? Where will this be shown? I know we were going to show google project id in the workspace (in case users need that information for gs commands via the terminal or something), but I am wondering where this will really be used, especially since our users won't have direct access.

Maybe we/support are the users of this?

Reply:
Janki and I went over the questions we had in the jira ticket and we decided to go with:

**display name**
* use the format “<namespace>--<name>” truncated to 30 chars

**Labels**
* use labels, truncating at 63 chars and replace all illegal characters with “-”

which should be reflected in the code now